include/boost/url/detail/impl/pct_encoding.ipp

```diff
@@ -81,6 +82,224 @@ key_equal_encoded(
     }
 }
 
+int
+pct_decode_compare_unchecked(
+    string_view lhs,
+    string_view rhs) noexcept
+{
+    auto consume_one = []
+        (string_view::iterator& it,
+         char &c,
+         std::size_t& n)
+    {
+        if(*it != '%')
+        {
+            c = *it;
+            ++it;
+        }
+        else
+        {
+            pct_decode_unchecked(
+                &c,
+                &c + 1,
+                string_view(it, 3));
+            it += 3;
+        }
+        ++n;
+    };
+
+    std::size_t n0 = 0;
+    std::size_t n1 = 0;
+    auto it0 = lhs.begin();
+    auto it1 = rhs.begin();
+    auto end0 = lhs.end();
+    auto end1 = rhs.end();
+    char c0 = 0;
+    char c1 = 0;
+    while (it0 < end0 &&
+           it1 < end1)
+    {
+        consume_one(it0, c0, n0);
+        consume_one(it1, c1, n1);
+        if (c0 < c1)
+            return -1;
+        if (c1 < c0)
+            return 1;
+    }
+    n0 += pct_decode_bytes_unchecked(
+        string_view(it0, end0 - it0));
+    n1 += pct_decode_bytes_unchecked(
+        string_view(it1, end1 - it1));
+    if (n0 == n1)
+        return 0;
+    if (n0 < n1)
+        return -1;
+    return 1;
+}
+
+std::size_t
+pct_decode_starts_with_unchecked(
+    string_view lhs,
+    string_view rhs) noexcept
+{
+    auto consume_one = []
+        (string_view::iterator& it,
+         char &c)
+    {
+        if(*it != '%')
+        {
+            c = *it;
+            ++it;
+            return;
+        }
+        pct_decode_unchecked(
+            &c,
+            &c + 1,
+            string_view(it, 3));
+        it += 3;
+    };
+
+    auto it0 = lhs.begin();
+    auto it1 = rhs.begin();
+    auto end0 = lhs.end();
+    auto end1 = rhs.end();
+    char c0 = 0;
+    char c1 = 0;
+    while (it0 < end0 &&
+           it1 < end1)
+    {
+        consume_one(it0, c0);
+        consume_one(it1, c1);
+        if (c0 != c1)
+            return 0;
+    }
+    if (it1 == end1)
+        return it0 - lhs.begin();
+    return 0;
+}
+
+std::size_t
+pct_decode_ends_with_unchecked(
```

Comment:
These functions are complicated enough that they deserve at least a sentence explaining what it does, as I cannot figure it out from the name alone, or the context.